// alt/lib/symbols/symbols.js
'use strict';

var _interopRequireWildcard = function (obj) { return obj && obj.__esModule ? obj : { 'default': obj }; };

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _Symbol = require('es-symbol');

var _Symbol2 = _interopRequireWildcard(_Symbol);

// action creator handler
var ACTION_HANDLER = _Symbol2['default']();

exports.ACTION_HANDLER = ACTION_HANDLER;
// the action's uid symbol for listening
var ACTION_KEY = _Symbol2['default']();

exports.ACTION_KEY = ACTION_KEY;
// per instance registry of actions
var ACTIONS_REGISTRY = _Symbol2['default']();

exports.ACTIONS_REGISTRY = ACTIONS_REGISTRY;
// the action's name
var ACTION_UID = _Symbol2['default']();

exports.ACTION_UID = ACTION_UID;
// store all of a store's listeners
var ALL_LISTENERS = _Symbol2['default']();

exports.ALL_LISTENERS = ALL_LISTENERS;
// initial snapshot
var INIT_SNAPSHOT = _Symbol2['default']();

exports.INIT_SNAPSHOT = INIT_SNAPSHOT;
// last snapshot
var LAST_SNAPSHOT = _Symbol2['default']();

exports.LAST_SNAPSHOT = LAST_SNAPSHOT;
// all lifecycle listeners
var LIFECYCLE = _Symbol2['default']();

exports.LIFECYCLE = LIFECYCLE;
// store action listeners
var LISTENERS = _Symbol2['default']();

exports.LISTENERS = LISTENERS;
// public methods
var PUBLIC_METHODS = _Symbol2['default']();

exports.PUBLIC_METHODS = PUBLIC_METHODS;
// contains all state
var STATE_CONTAINER = _Symbol2['default']();
exports.STATE_CONTAINER = STATE_CONTAINER;
