// alt/utils/AltTestingUtils.js
'use strict';

var _interopRequireWildcard = function (obj) { return obj && obj.__esModule ? obj : { 'default': obj }; };

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } };

var _get = function get(object, property, receiver) { var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _inherits = function (subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; };

var _assign = require('object-assign');

var _assign2 = _interopRequireWildcard(_assign);

var noop = function noop() {};

var AltTestingUtils = {
  createStoreSpy: function createStoreSpy(alt) {
    return {
      _storeName: 'ALT_TEST_STORE',
      alt: alt,
      bindAction: noop,
      bindActions: noop,
      bindListeners: noop,
      dispatcher: alt.dispatcher,
      emitChange: noop,
      exportPublicMethods: noop,
      getInstance: noop,
      on: noop,
      setState: noop,
      waitFor: noop
    };
  },

  makeStoreTestable: function makeStoreTestable(alt, UnwrappedStore) {
    var StorePrototype = AltTestingUtils.createStoreSpy(alt);

    var DerivedStore = (function (_UnwrappedStore) {
      function DerivedStore() {
        _classCallCheck(this, DerivedStore);

        _get(Object.getPrototypeOf(DerivedStore.prototype), 'constructor', this).call(this);
      }

      _inherits(DerivedStore, _UnwrappedStore);

      return DerivedStore;
    })(UnwrappedStore);

    _assign2['default'](DerivedStore.prototype, StorePrototype);
    return new DerivedStore();
  },

  mockGetState: function mockGetState() {
    var state = arguments[0] === undefined ? {} : arguments[0];

    return {
      getState: function getState() {
        return state;
      }
    };
  }
};

module.exports = AltTestingUtils;
