// alt/lib/addons.js
'use strict';

var _interopRequireWildcard = function (obj) { return obj && obj.__esModule ? obj : { 'default': obj }; };

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _Alt = require('./');

var _Alt2 = _interopRequireWildcard(_Alt);

var _ActionListeners = require('../utils/ActionListeners');

var _ActionListeners2 = _interopRequireWildcard(_ActionListeners);

var _AltManager = require('../utils/AltManager');

var _AltManager2 = _interopRequireWildcard(_AltManager);

var _DispatcherRecorder = require('../utils/DispatcherRecorder');

var _DispatcherRecorder2 = _interopRequireWildcard(_DispatcherRecorder);

var _atomic = require('../utils/atomic');

var _atomic2 = _interopRequireWildcard(_atomic);

var _connectToStores = require('../utils/connectToStores');

var _connectToStores2 = _interopRequireWildcard(_connectToStores);

var _chromeDebug = require('../utils/chromeDebug');

var _chromeDebug2 = _interopRequireWildcard(_chromeDebug);

var _makeFinalStore = require('../utils/makeFinalStore');

var _makeFinalStore2 = _interopRequireWildcard(_makeFinalStore);

var _withAltContext = require('../utils/withAltContext');

var _withAltContext2 = _interopRequireWildcard(_withAltContext);

var _AltContainer = require('../../AltContainer');

var _AltContainer2 = _interopRequireWildcard(_AltContainer);

_Alt2['default'].addons = {
  ActionListeners: _ActionListeners2['default'],
  AltContainer: _AltContainer2['default'],
  AltManager: _AltManager2['default'],
  DispatcherRecorder: _DispatcherRecorder2['default'],
  atomic: _atomic2['default'],
  chromeDebug: _chromeDebug2['default'],
  connectToStores: _connectToStores2['default'],
  makeFinalStore: _makeFinalStore2['default'],
  withAltContext: _withAltContext2['default']
};

exports['default'] = _Alt2['default'];
module.exports = exports['default'];
